

## Existing Reference
Statistics regarding the provided reference genome

### Basic stats

```{r importRefGenome, warning=FALSE, results="asis"}
if(refAvail){
  refGenome.file <- getConfigField("genome")
  if(substr(refGenome.file,1,1)!="/") refGenome.file <- file.path(projFolder, refGenome.file)
  refGenome <- importFA(refGenome.file)
}
```

```{r show ref stats}
if(refAvail){
  out <- summary(refGenome)
  out[1,1] <- c("No. of Chromosomes/Contigs")
  out[,2] <- formatC(out[,2],  format="d", big.mark=",")
  out_html <- knitr::kable(out, col.names = NULL, "html")
  
  kable_styling(out_html, "striped", position = "left")
}
```


## Mock vs reference

These are the flagstats from mapping the mock clusters against an existing reference genome. Typical flag stats are


|     Flag | Explanation                          |
|---------:|-------------------------------------:|
|        0 | Read aligned against forward strand  |
|       16 | Read aligned against reverse strand  |
|     2048 | Supplementary alignment forward      |
|     2064 | Supplementary alignment reverse      |
|        4 | Unmapped                             |


```{r samflags}
if(refAvail){
samFlagstats <- read.table(file.path(projFolder,"BAM", "Mockref", "mockToRef.sam.samflags"))
colnames(samFlagstats) <- c("Freq.", "Flag")
totalLoci <- samFlagstats$V1[samFlagstats$V2==0] + samFlagstats$V1[samFlagstats$V2==16]
out_html <- knitr::kable(samFlagstats, "html")
kable_styling(out_html, "striped", position = "left")
} else {
  totalLoci <- 0
}
```

From there we have `r formatC(totalLoci,  format="d", big.mark=",")` primary alignments.

```{r bam.merged.bed}
if(refAvail){
unmergedLoci <- read.table(file.path(projFolder,"BAM","Mockref","mockToRef.bed"))
mergedLoci <- read.table(file.path(projFolder,"BAM","Mockref","mockToRef.merged.bed"))
totalBases <- sum(abs(mergedLoci$V3 - mergedLoci$V2))
} else {
  unmergedLoci <- 0
  mergedLoci <- 0
  totalBases <- 0
}
```

However, in total we have `r formatC(nrow(unmergedLoci),  format="d", big.mark=",")` mapping loci that can be merged down (by intersection of the bed) to `r formatC(nrow(mergedLoci),  format="d", big.mark=",")` if their chromosomal location is intersected.
Please note, if the difference between these two numbers is too large, it indicates that the mock reference could possibly be merged more strictly.

In total, `r formatC(totalBases, big.mark=",")` bases are covered in the reference genome from mock cluster sequences.

### Alignments per Chromosome in Reference

A visualisation of the alignments of the mock clusters to the reference genome

```{r}
if(refAvail){
  par(oma=c(8,1,1,0))
mappedLoci <- read.table(file.path(projFolder,"BAM","Mockref","mockToRef.coverage"))
barplot(mappedLoci[,2], main="Alignments per Chromosome in Reference", col=report.colours[1], names=mappedLoci[,1],las=2)
}
```

```{r SAMbed}
if(refAvail){
 SAMLoci <- read.table(file.path(projFolder,"BAM","Mockref","mockToRef.bed"))
}
```

This is the length distribution of the sequences of mock reference clusters mapped against the reference genome

```{r}
if(refAvail){
  out <- t(as.data.frame(as.matrix((summary(SAMLoci$V3-SAMLoci$V2)))))
  rownames(out) <- "5-point summary"
  out_html <- knitr::kable(t(out), "html")
  kable_styling(out_html, "striped", position = "left")  
}
```

### Avg cluster mapping distance per Chromosome

Then, this is the average distance from one alignment to the next, per chromosome

```{r}
if(refAvail){
chrNames <- unique(SAMLoci$V1)
avgDist <- c()
for(i in 1:length(chrNames)){
  tmp <- SAMLoci[SAMLoci$V1==chrNames[i],]
  avgDist[i] <- mean(abs(tmp$V2[1:(nrow(tmp)-1)] - tmp$V2[2:(nrow(tmp))]))
}
}
```

```{r}
if(refAvail){
barplot(avgDist, main="Avg cluster mapping distance per Chromosome", col=report.colours[1], names=chrNames, las=2)
}
```

### Mock reference contig alignments on reference genome

Coverage of the lined up alignment loci of the mock reference contigs against the reference genome. Ideally, we observe here just a constant coverage of 1, meaning that each contig of the mock reference aligns to a unique place on the reference genome. Having here spikes indicates that several mock reference contigs align to one and the same loci along the reference genome and that those contigs could possibly be merged.

```{r}
if(refAvail){
  barplot(SAMLoci$V4, main="Mock reference contig alignments on reference genome")
}
```

Here is a list of the most covered areas (coverage > 10)

```{r}

tmp <- SAMLoci[order(SAMLoci[,4], decreasing = TRUE),]
tmp <- tmp[tmp[,4]>10,]

if(nrow(tmp)>0){
  rownames(tmp) <- 1:nrow(tmp)
  colnames(tmp) <- c("Chromosome", "Start", "End", "Coverage")
  #knitr::kable(tmp) %>% kable_styling
  datatable(tmp, extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                           lengthMenu = list(c(10,25,50,-1),
                                             c(10,25,50,"All"))))
}
``` 

In summary, we have the following quantities of alignments of mock reference contigs against various loci on the reference genome. Value 1 means here, only a single mock reference cluster aligns to a particular reference genome locus, two means, there are two different mock reference clusters that align to a reference genome locu and so forth.

```{r summaryTable cluster to ref genome, results = 'asis'}

  out <- table(SAMLoci$V4)
# Split the table
  printRows <- length(out)%/%10  
  for(i in 1:printRows){
    out_html <- knitr::kable(t(out[((i-1)*10+1):(i*10)]), "html")
    print(kable_styling(out_html, "striped", position = "left") ) 
  }
```


```{r}
if(refAvail){
flankingCounts <- read.table(file.path(projFolder,"BAM","Mockref","mockToRef.merged.combined.fa.counts"))
} else {
  flankingCounts <- 0
}
```

In total we have `r nrow(flankingCounts)` many different flanking sequence combinations (+-6 bases), here is the top 20 for the unmerged set of sequences.

```{r}
if(refAvail){
tmp <- flankingCounts[order(flankingCounts$V1, decreasing=TRUE),][1:20,]
rownames(tmp) <- NULL
colnames(tmp) <- c("Counts", "Sequence")
#  out_html <- knitr::kable(tmp, "html")
#  kable_styling(out_html, "striped", position = "left")  
if(nrow(tmp)>0){
  rownames(tmp) <- 1:nrow(tmp)
  colnames(tmp) <- c("Counts", "Sequence")
  #knitr::kable(tmp) %>% kable_styling
  datatable(tmp, extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                           lengthMenu = list(c(10,25,50,-1),
                                             c(10,25,50,"All"))))
}
}
```

```{r}
if(refAvail){
mergedFlankingCounts <- read.table(file.path(projFolder,"BAM", "Mockref","mockToRef.merged.combined.fa.counts"))
} else {
  mergedFlankingCounts <- 0
}
```

In total we have `r nrow(mergedFlankingCounts)` many different flanking sequence combinations (+-6 bases) in the merged set, here is the top 20 for the unmerged set of sequences.

```{r}
if(refAvail){
tmp <- mergedFlankingCounts[order(mergedFlankingCounts$V1, decreasing=TRUE),][1:20,]
rownames(tmp) <- NULL
colnames(tmp) <- c("Counts", "Sequence")
if(nrow(tmp)>0){
  rownames(tmp) <- 1:nrow(tmp)
  colnames(tmp) <- c("Counts", "Sequence")
  #knitr::kable(tmp) %>% kable_styling
  datatable(tmp, extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                           lengthMenu = list(c(10,25,50,-1),
                                             c(10,25,50,"All"))))
}
}
```

