
# Homology comparison

## Visualization of VCF
Here, we have Homo. Ref green, Homo. Alt red, hetero blue and missing black

```{r image vcf, fig.width=15, fig.height=10, warning=FALSE}
par(mar=c(10,5,1,1))
plotThis <- t(apply(as.matrix(vcf$genotypes),1,as.numeric))
dataForBarplot <- apply(plotThis,1,class.table, classes=c(0,1,2,3))
colnames(dataForBarplot) <- rownames(vcf$genotypes)
barplot(dataForBarplot, col=c("green", "blue", "red", "black"), las=2)
```

## PCA

```{r, warning=FALSE}
 vcf2 <- importVCF(file.path(projFolder, "VCF", "FinalSetVariants_finalMock.vcf"))
#vcf2 <- importVCF(file.path(projFolder, "MPILEUP", "mpileup_reference", "GSC.vcf"))

tmp <- as.matrix(vcf2$genotypes)
tmp <- apply(tmp,1,as.numeric)
tmp[is.na(tmp)] <- 0

vcf.PCA <- prcomp(t(tmp))
```

```{r}
palette("default")
pairs(vcf.PCA$x[,1:3])

if(!is.null(sampleInfo)){
for(i in 2:ncol(sampleInfo)){
  pairs(vcf.PCA$x[,1:3], col=as.factor(sampleInfo[,i]), pch=20)
}
}
```

```{r}
cols <- min(ncol(vcf.PCA$x),10)
pairs(vcf.PCA$x[,1:cols])

if(!is.null(sampleInfo)){
for(i in 2:ncol(sampleInfo)){
  pairs(vcf.PCA$x[,1:cols], col=as.numeric(as.factor(sampleInfo[,i])), pch=20)
}
}
```

## Other methods

### t-SNE
```{r}
cols <- min(ncol(vcf.PCA$x),10)
TSNEout <- tsne(vcf.PCA$x[,1:cols], k=5)

if(!is.null(sampleInfo)){
for(i in 2:ncol(sampleInfo)){
  plot(TSNEout, col=as.numeric(as.factor(sampleInfo[,i])), pch=20)
}
} else {
   plot(TSNEout, pch=20)
}
```

### Kernel PCA
```{r}
# Kernel PCA
cols <- min(min(dim(vcf.PCA$x)),10)-1

KPCAres <- kpca(vcf.PCA$x[,1:cols])

if(!is.null(sampleInfo)){
  for(i in 2:ncol(sampleInfo)){
     pairs(rotated(KPCAres)[,1:cols], col=as.numeric(as.factor(sampleInfo[,i])), pch=20)
  }

} else {
   ndim <- min(ncol(rotated(KPCAres)), 10)
   plot(rotated(KPCAres)[,1:ndim], pch=20)
}

```

