# Variants
Import the Result vcf

```{r import vcf, echo=FALSE, message=FALSE, warning=FALSE}
vcf <- importVCF(file.path(projFolder, "MPILEUP", "mpileup_finalMock", "GSC.vcf"))
vcfUnfiltered <- importVCF(file.path(projFolder, "MPILEUP", "mpileup_finalMock", "GSC.unfiltered.vcf"))
if(refAvail){
  vcfRef <- importVCF(file.path(projFolder, "VCF", "FinalSetVariants_referenceGenome.vcf"))
}
```

```{r get alleleCounts}
gtInfo <- apply(vcfUnfiltered$genotypesInfo,2,strsplit,",")
refAll <- matrix(-1, nrow=length(gtInfo), ncol=length(gtInfo[[1]]))
altAll <- matrix(-1, nrow=length(gtInfo), ncol=length(gtInfo[[1]]))

for(i in 1:nrow(refAll)){
  tmp <- sapply(gtInfo[[i]],"[",1)
  tmp[tmp=="."] <- NA
  refAll[i,] <- as.numeric(tmp)
  
  tmp <- sapply(gtInfo[[i]],"[",2)
  tmp[tmp=="."] <- NA
  altAll[i,] <- as.numeric(tmp)
}

DP <- refAll+altAll
```

## Mock reference
We give some basic stats on the vcf file

```{r}
  out <- c()
  out[1] <- dim(vcf$map)[1]
  noVariants <- dim(vcf$map)[1]
  names(out)[1] <- "Total number of variants"
  out[2] <- sum(sapply(apply(vcf$genotypes,2,table),length)==1)
  names(out)[2] <- "Monomorphic sites"
  out_html <- knitr::kable(out, col.names = NULL, "html")
  kable_styling(out_html, "striped", position = "left")
```

### SNP position vs snp index
Plot the snp position

```{r}
plot(as.vector(as.matrix(vcf$map[,4])), ylab="Pos. on mock", xlab="SNP index")
```

### SNP loci on cluster
Distribution of the variant across the cluster

```{r}
hist(as.vector(as.matrix(vcf$map[,4])), main="SNP loci distribution", xlab="Position", col=report.colours[1])
```

### Missing variants per samples

Missing variant calls per sample

```{r}
missingnessPerSample <- apply(vcf$genotypesInfo==".,.",2,sum)
barplot(missingnessPerSample, las=2, main="Missing variants per sample", col=report.colours[1])
missingnessPerSample <- as.matrix(missingnessPerSample)
```

### Missigness vs sequencing depths

```{r}
missSample <- data.frame(Sample=gsub("\\.","_",rownames(missingnessPerSample)),
                         Missing=missingnessPerSample)
missVSseq <- merge(totalSequences, missSample, by="Sample")
plot(missVSseq[,3], missVSseq[,2])
```

### Missingness per bp position
t.b.a.

### Missing variants per samples

```{r}
barplot(apply(vcf$genotypesInfo==".,.",2,sum)/noVariants*100, las=2, main="Missing variants per sample (in %)", col=report.colours[1])
```

### Distribution of filtering criterias
```{r}
tmp <- cumsum(sort(apply(DP,2,sum, na.rm=TRUE)))
plot(tmp, type="l")
abline(a=0, b=3*70, col="red")
abline(a=0, b=200*70, col="red")
```

```{r}
tmp <- table(apply(DP>200,2,sum, na.rm=TRUE))
tmp <- rbind(tmp,table(apply(DP>2,2,sum)))
tmp <- rbind(tmp,table(apply(DP>3,2,sum)))
tmp <- rbind(tmp,table(apply(DP>4,2,sum)))
tmp <- rbind(tmp,table(apply(DP>5,2,sum)))
barplot(tmp,border=0)
```


## Reference genome
We give some basic stats on the vcf file

```{r}
if(refAvail){
  out <- c()
  out[1] <- dim(vcfRef$map)[1]
  noVariants.ref <- dim(vcfRef$map)[1]
  names(out)[1] <- "Total number of variants"
  out[2] <- sum(sapply(apply(vcfRef$genotypes,2,table),length)==1)
  names(out)[2] <- "Monomorphic sites"
  out_html <- knitr::kable(out, col.names = NULL, "html")
  kable_styling(out_html, "striped", position = "left")
}
```

Plot the snp position
```{r}
if(refAvail){
  plot(as.vector(as.matrix(vcfRef$map[,4])), ylab="Pos. on mock", xlab="SNP index")
}
```
Missing variant calls per sample

```{r}
if(refAvail){
  barplot(apply(vcfRef$genotypesInfo==".,.",2,sum), las=2, main="Missing variants per sample", col=report.colours[1])
}
```
```{r}
if(refAvail){
  barplot(apply(vcfRef$genotypesInfo==".,.",2,sum)/noVariants.ref*100, las=2, main="Missing variants per sample (in %)", col=report.colours[1])
}
```

